axlearn/common/checkpointer.py

```diff
@@ -922,3 +971,216 @@ def validate_and_restore(*, step: int, ckpt_dir: str):
+            else:
+                spec["index"].append((path, value))
+        return spec
+
+    # pylint: disable-next=redefined-builtin
+    def ckpt_dir(self, step: int, dir: Optional[str] = None) -> str:
+        """Obtains the checkpoint dir for the given step."""
+        if dir is None:
+            dir = self._manager.directory
+        # pylint: disable-next=protected-access
+        return str(self._manager._get_save_directory(step, dir))
+
+    def save(
+        self, *, step: int, state: Nested[Tensor], evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """See `BaseCheckpointer.save` for details.
+
+        Checkpoint saving is handled by `orbax` checkpoint manager.
+        """
+        import orbax.checkpoint as ocp
+
+        spec = self._get_spec(step=step, state=state)
+        self._eval_summaries_sem.acquire()  # pylint: disable=consider-using-with
+        assert not self._eval_summaries
+        self._eval_summaries.update(evaler_summaries or {})
+        self._manager.save(
+            step=step,
+            # The input iterator is saved as part of `save_tf_savables`.
+            args=ocp.args.Composite(
+                index=self._index_args(spec["index"]),
+                tf_ckpt_map=self._tf_ckpt_args(spec["tf_ckpt_map"]),
+                state=ocp.args.StandardSave(
+                    utils.prune_tree(
+                        state, should_prune=lambda _, v: isinstance(v, tf.data.Iterator)
+                    )
+                ),
+            ),
+        )
+
+    def restore(
+        self,
+        *,
+        step: Optional[int] = None,
+        state: Union[Nested[Tensor], Nested[TensorSpec]],
+    ) -> Tuple[Optional[int], Nested[Tensor]]:
+        """See `BaseCheckpointer.restore` for details."""
+        import orbax.checkpoint as ocp
+
+        cfg: OrbaxCheckpointer.Config = self.config
+
+        spec = self._get_spec(step=step, state=state)
+        state_specs = jax.tree_util.tree_map(
+            lambda x: jax.ShapeDtypeStruct(shape=x.shape, dtype=x.dtype, sharding=x.sharding),
+            jax.tree_util.tree_map(lambda x: None if isinstance(x, tf.data.Iterator) else x, state),
+        )
+        try:
+            composite_state = self._manager.restore(
+                step,
+                args=ocp.args.Composite(
+                    tf_ckpt_map=self._tf_ckpt_args(spec["tf_ckpt_map"]),
+                    state=ocp.args.StandardRestore(state_specs),
+                ),
+            )
+            restored_state = composite_state["state"]
+        except Exception as e:  # pylint: disable=broad-except
+            if step is not None:
+                raise ValueError(f"Failed to restore at step {step}.") from e
+            logging.info("Could not find any completed checkpoints under %s: %s", cfg.dir, e)
+            restored_state = state
+
+        # If we successfully restored from step=None, use the latest step.
+        if step is None:
```

Comment:
My preference would be for doing this check first and raising an error earlier if no steps are found, followed by the try/except. I think it's a bit more readable but up to you.

Reply:
Thanks -- the existing behavior is to use the input state when restoring from a checkpoint dir without checkpoints, but raise if step is provided explicitly (so in the except we want to distinguish None vs not-None).